### PR TITLE
MDEV-39169: Replace deprecated network functions with modern POSIX equivalents

### PR DESCRIPTION
**Pull request by FaramosCZ**
resolveip: replace inet_aton/inet_addr, gethostbyname, gethostbyaddr, inet_ntoa with inet_pton, getaddrinfo, getnameinfo, inet_ntop.

These deprecated functions are IPv4-only and some are not thread-safe (gethostbyname, gethostbyaddr, and inet_ntoa use static internal buffers). The modernized code:

- Adds IPv6 support for both forward and reverse lookups
- Uses thread-safe APIs throughout
- Eliminates the #ifdef WIN32 code path split (inet_pton works on both)
- Removes the h_errno dependency (replaced by gai_strerror)
- Uses AF_UNSPEC with getaddrinfo to return both IPv4 and IPv6 results
- Preserves existing output format, exit codes, and special cases (Broadcast, Null-IP-Addr, Old-Bcast)

Bump version from 2.3 to 2.4 to reflect the API change.

Update Solaris library search in CMakeLists.txt from inet_aton to getaddrinfo since the old function is no longer referenced.

### CODE REVIEW

**Review by gkodinov — APPROVED** [review]
Thank you for your contribution! This is a preliminary review.

LGTM. I've taken the liberty of filing a jira for the task. 

Please stand by for the final review. 

I'd also be nice to have some sort of testing for this. Maybe resolve some IPv6 IP or something?

File: mysql-test/main/resolveip.test [hunk 1]
```diff
@@ -0,0 +1,37 @@
+--echo #
+--echo # MDEV-39169: resolveip IPv6 support
+--echo #
+
+--source include/not_windows.inc
```
Comment (raghunandanbhat) [inline mysql-test/main/resolveip.test]: please move the `--source include..` to the top of the test file.

we write our mtr test in this format - 
```
--source include/....inc 

--echo # 
--echo # MDEV-39169: resolveip IPv6 support
--echo # 

... <tests> ...

--echo # End of 13.x tests 
```
you have the header with MDEV, but footer is missing. please add the footer in your test with correct server version.

File: mysql-test/main/resolveip.test [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+--echo #
+--echo # IPv4 reverse lookup (silent mode)
+--echo #
+--exec $MYSQL_RESOLVEIP -s 127.0.0.1 | sed 's/.*/HOSTNAME/'
+
+--echo #
+--echo # IPv6 reverse lookup
+--echo #
+--exec $MYSQL_RESOLVEIP ::1 | sed 's/Host name of ::1 is .*/Host name of ::1 is HOSTNAME/'
```
Comment (raghunandanbhat) [inline mysql-test/main/resolveip.test]: these are some of the cases that were not resolved in my machine (Ubuntu 24) with resolveip, I believe these are valid IPv6 addresses
```
$ resolveip :: ::1 ::ffff:127.0.0.1 0000:0000:0000:0000:0000:0000:0000:0001
resolveip: Unable to find hostid for '::': host not found
resolveip: Unable to find hostid for '::1': host not found
resolveip: Unable to find hostid for '::ffff:127.0.0.1': host not found
resolveip: Unable to find hostid for '0000:0000:0000:0000:0000:0000:0000:0001': host not found
```
`getent ahosts` works but resolveip couldn't resolve them 
```
$ getent ahosts :: ::1 ::ffff:127.0.0.1 0000:0000:0000:0000:0000:0000:0000:0001
::              STREAM ::
::              DGRAM  
::              RAW    
::1             STREAM ::1
::1             DGRAM  
::1             RAW    
::ffff:127.0.0.1 STREAM ::ffff:127.0.0.1
::ffff:127.0.0.1 DGRAM  
::ffff:127.0.0.1 RAW    
::1             STREAM 0000:0000:0000:0000:0000:0000:0000:0001
::1             DGRAM  
::1             RAW  
```

File: mysql-test/main/resolveip.test [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+--echo #
+--echo # IPv6 reverse lookup (silent mode)
+--echo #
+--exec $MYSQL_RESOLVEIP -s ::1 | sed 's/.*/HOSTNAME/'
+
+--echo #
+--echo # Error: invalid hostname
```
Comment (raghunandanbhat) [inline mysql-test/main/resolveip.test]: for completeness, may be you can add few more tests- trying to resolve multiple addresses, invalid addresses, resolving multiple mixed args ( `resolveip ipv4 ipv6`, `resolveip ipv4 hostname`)

**Review by raghunandanbhat — COMMENTED** [review]
@FaramosCZ , thanks for the contribution. 

I have a few suggestion-
- please squash two commits into a single commit
- ensure that commit message starts with the MDEV number, something like `MDEV-39169:...`
- plus, issues I found in test (comments below).

**Comment by FaramosCZ** [comment]
I fixed all concerns you listed.
Let me know if anything else needs changing.

File: mysql-test/main/resolveip.test [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+--echo #
+--echo # IPv4 special addresses (deterministic, no DNS needed)
+--echo #
+--exec $MYSQL_RESOLVEIP 255.255.255.255
+--exec $MYSQL_RESOLVEIP 0.0.0.0
+
+--echo #
+--echo # IPv4 reverse lookup
+--echo #
+--exec $MYSQL_RESOLVEIP 127.0.0.1 2>&1 | sed 's/Host name of 127\.0\.0\.1 is .*/Host name of 127.0.0.1 is HOSTNAME/' | sed "s|.*Unable to find hostname for '127.0.0.1'|Host name of 127.0.0.1 is HOSTNAME|"
```
Comment (raghunandanbhat) [inline mysql-test/main/resolveip.test]: `sed` here is masking a failure by transforming an error into a success string. This makes the test deceptive; it will "pass" even if the utility fails to resolve the IP.
specifically this part-
```
sed "s|.*Unable to find hostname for '127.0.0.1'|Host name of 127.0.0.1 is HOSTNAME|"
```

also present in other tests - resolving `::1`, `::ffff:127.0.0.1`, `0000:0000:0000:0000:0000:0000:0000:0001`

**Review by raghunandanbhat — CHANGES_REQUESTED** [review]
@FaramosCZ, looks good overall. 

cases like `::`, `::1`, IPv4-mapped IPv6 addresses, etc seem to work fine. however, some test cases are "sneaky". please fix them.

File: mysql-test/main/resolveip.test (hunk 4 above)
Comment (FaramosCZ) [reply mysql-test/main/resolveip.test]: Right. I removed the error-masking sed, leaving just the hostname-normalizing sed.

The original idea behind the masking was to make it system agnostic.

Reverse DNS resolution via `getnameinfo()` with `NI_NAMEREQD` is system-dependent — it relies on `/etc/hosts` entries (or functioning DNS) for the tested addresses. Unlike forward lookups, there is no MTR pre-check for reverse DNS availability. The existing `check_ipv6.inc` only verifies that the kernel can create and bind an `AF_INET6` socket (via suite.pm's `ipv6_ok()`); it says nothing about name resolution.


There is no easy way to cover that in the test.
I added an extra comment at the top of the test for the future person debugging it in an unusual environments.

For current Fedora / Debian / Ubuntu BuildBot targets, this doesn't seem to be a problem.

---

On several places there is `sed 's/.*/HOSTNAME/'`, that replaces the complete output, but these are the silent-mode and mixed-args tests. This should be OK since silent mode outputs a bare hostname that's unpredictable.

**Review by raghunandanbhat — DISMISSED** [review]
looks good to me. thank you @FaramosCZ  for the contribution.

@gkodinov, I believe this needs go through QA/testing before we merge this.

File: mysql-test/main/resolveip.test (hunk 4 above)
Comment (raghunandanbhat) [reply mysql-test/main/resolveip.test]: I suggest, we can create a check similar to `check_ipv6.inc` in a new .inc file like this -
```
--error 0,1,2
--exec getent hosts 127.0.0.1 ::1 ::ffff:127.0.0.1 > /dev/null 2>&1
if ($errno)
{
    skip IPv4-mapped IPv6 address resolution not possible;
}
```
and move the tests that rely on having entries in `/etc/hosts` to a new file.  skip test if the check fails. 

what do you think? would this fix the problem?

Comment (FaramosCZ) [reply mysql-test/main/resolveip.test]: It solves part of the problem, but adds a new ones.

- `getent hosts <IP>` uses NSS `gethostbyaddr()` internally, which typically consults the same `/etc/hosts` but goes through a different `glibc` code path and can behave differently under custom NSS configurations (/`etc/nsswitch.conf`). On exotic NSS setups (LDAP, SSSD, custom modules), they could diverge — the check would pass but getnameinfo fails, or vice versa.

- `getent` doesn't exist on macOS
While not in BB, I see there is some [MariaDB server on macOS](https://mariadb.com/docs/server/server-management/install-and-upgrade-mariadb/installing-mariadb/binary-packages/installing-mariadb-on-macos-using-homebrew)
`Command not found` would return shell exit code `127`

- `getent` hosts checks all three addresses as a batch. Return success only if all resolves.

I think it's fine to just ignore the uncommon setups in this case. We could spend unreasonable effort fixing problems that potentially no one will have and I'm sure it will be always possible to invent some new exotic corner cases if someone would want to, for us to cover.

I believe the comment I added is a solid hint on what might went wrong and why, for anyone actually bothered by the test fail. And on those setups, the person creating the uncommon networking setup should be smart enough to figure out some solution.
